### ISSUE TITLE
More support for deprecated fields/models

### ISSUE THREAD

**Issue by Viicos** [pydantic/pydantic#8922]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

https://github.com/pydantic/pydantic/pull/8237 implemented support for deprecated fields. While doing so, we encountered/postponed some extra features that could be nice to have:

1. When using the [`deprecated`](https://docs.python.org/3.13/library/warnings.html#warnings.deprecated) decorator class, extra information (such as the `category` and `stacklevel`) can be passed. It would be great if Pydantic could support such attributes when emitting the runtime warning. For reference, this commit implements the feature: https://github.com/pydantic/pydantic/pull/8237/commits/cf06a7322d59aefc2e3ca607f0c5ee66d28f6757 (extra care should be taken with the `typing-extensions` version -- Only 4.9.0 and greater support the class implementation of the `deprecated` decorator).

2. When instantiating a model with a deprecated field explicitly specified (`Model(deprecated_field=...)`), the runtime warning could be emitted.

3. Support for Pydantic/stdlib dataclasses: not sure if there are any technical limitations for this one?

4. Support for deprecated models:
  - With the `@deprecated` decorator:
    ```python
    @deprecated("deprecated model")
    class Model(BaseModel):
        pass
    ```
    The decorator takes care of the runtime warning (tested locally and doesn't seem to cause any issues with the model metaclass). However, the JSON Schema should reflect the deprecation.
  - via a config option, in a similar fashion to the argument to `Field`:
    ```python
    class Model(BaseModel):
        model_config = {"deprecated": "deprecated model"}
    ```
    I wouldn't really recommend the second option, as type checkers won't be aware of the deprecation (same for `computed_field`)

These were the ideas I had while implementing support for deprecated fields. Some of them requires some discussion, as I'm not sure whether it should be supported or not. Happy to hear your opinions

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [X] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by sydney-runkle** [comment]
Agreed. PRs welcome for parts or all of the above features. Ping me if you want to chat more about implementation details :).

**Comment by NeevCohen** [comment]
@sydney-runkle @Viicos I had started to look into implementing point 4, and from my testing, the `@deprecated` decorator on a model does emit a runtime warning when using the model.

My question is, should a warning also be emitted in the following scenario? 

```python
@deprecated
class DeprecatedModel(BaseModel):
    pass

class NewModel(BaseModel):
    field: DeprecatedModel
```

Do you think that in this case Pydantic should emit a warning along the lines of `` Usage of deprecated model `DeprecatedModel` ``?

**Comment by Viicos** [comment]
That's great to hear!

> Do you think that in this case Pydantic should emit a warning along the lines of `` Usage of deprecated model `DeprecatedModel` ``?

Do you mean the warning should be emitted during class creation? (And thus whenever the model's module is imported).
That might be too invasive, although I can't think of a use case where a user would be constrained to use a deprecated model as a field in the model he is defining. Let's maybe here what the maintainers think of it

**Comment by NeevCohen** [comment]
> Do you mean the warning should be emitted during class creation? (And thus whenever the model's module is imported). That might be too invasive, although I can't think of a use case where a user would be constrained to use a deprecated model as a field in the model he is defining. Let's maybe here what the maintainers think of it
 
I actually meant to emit a warning when a user _defines_ a model with one of its fields being a deprecated model, not when the deprecated model is created/imported.

In my examples above, when creating `NewModel`, we could check to see if a field is defined with the type of another subclass of `BaseModel` and then check if has the `__deprecated__` attribute.

**Comment by sydney-runkle** [comment]
Also, just a side note, I don't think we should support the config based deprecation above, but the deprecated decorator approach is great.

**Comment by allenlawrence94** [comment]
Has any progress been made on:
> When instantiating a model with a deprecated field explicitly specified (Model(deprecated_field=...)), the runtime warning could be emitted.

I'm currently doing this with a validator, but this would be nice to have.

**Comment by sydney-runkle** [comment]
Nope, not yet @allenlawrence94. Feel free to open a PR with support this!

**Comment by changhc** [comment]
I'd like to get some comments here for David's comment: https://github.com/pydantic/pydantic-core/pull/1551#issuecomment-2505791080

There is some concern about this new deprecation warning breaking users' existing code. To address this, I think we can add one new model config like `suppress_deprecation_warning` which allows users to ignore warnings due to deprecated fields being filled or accessed. Essentially, this new model config is equivalent to disabling/nullifying any support for deprecated fields because the only difference between deprecated fields and regular fields is the warning. One important thing here, in my opinion, is to either keep all warnings or raise nothing at all. We shouldn't allow users to partially suppress warnings, e.g. "no warning at model instantiation please but we want warnings when fields are accessed." I don't think there's any use case that really makes sense -- it's kind of strange to me that one needs such specific control for deprecation warning.

Meanwhile, unlike what David mentioned in his comment, I think this "suppress_deprecation_warning" thing should be opt-out rather than opt-in. This being opt-in basically means turning off support for deprecation fields by default and that this feature needs one additional step to be enabled. For instance, if one wants to use deprecated fields, this user needs to not only annotate those fields as deprecated but also set something like `enable_deprecated_fields` in model config, which is like double opting in. This is awkward and definitely leads to bad user experience in my opinion. Also, since the support for deprecation fields is already released, making this whole warning suppression thing opt-in will possibly break some other users' code as well. On the other hand, if we go for opt-out, some users' code might break after they upgrade pydantic, but they can then set this model config to opt out so that they can either completely ignore these errors or buy some time to fix the warnings. As what I said above, I think it's unlikely that users are okay with deprecation warnings at some specific spots only, so eventually they will probably just fix the warnings or handle them somehow and remove the model config. Another thing about making this opt-out by default is that this matches what one gets with the `@deprecated` decorator, e.g. deprecated models -- users opt in and get warnings from whatever is decorated. There isn't even an option for opting out, which is natural (why opt out? Either remove references to the model or remove the decorator!)

Overall, I think we are thinking too much about this warning breaking anything. If a deprecation warning does break something, that thing should have been handled properly. We can provide users with an option to suppress such warnings, but that should be a temporary solution.

Please let me know what you think. Thanks!

**Comment by Viicos** [comment]
> e.g. "no warning at model instantiation please but we want warnings when fields are accessed." I don't think there's any use case that really makes sense -- it's kind of strange to me that one needs such specific control for deprecation warning.

Actually, I think there is. It all depends on _who_ is consuming/using the model. If as a library, you provide instances of a model you control (say consumers of the library make use of a model as an annotation in a FastAPI route, and the route function implementation makes use of the model fields), you want to warn on attribute access, but you may not want the warning to be raised when you instantiate the model (because you are the "owner", you control the fields being defined, so you know you can safely instantiate a model with a deprecated field, and when you change the model definition, you are a able to not provide the field value anymore).

It also depends on the reason why the field is deprecated. I can see two scenarios for a field to be deprecated:
- replaced by a new one. In that case, it would make sense to warn on instantiation, as you can just provide the value for the new field.
- planned for deletion. In that case, you will still have to provide a value, the deprecation warning is just here to warn consumers of the model.

Instead of introducing a new config value, we were thinking about introducing a new `Deprecation` helper class, so that you can use it this way:

```python
class Model(BaseModel):
    a: int = Field(deprecated=Deprecation("'a' is deprecated", mode='warn-on-instantiation'))
```

But this gets pretty complex.

**Comment by changhc** [comment]
I see. You are right. I only thought about the cases where users instantiate models and consumes the instances and completely missed out common uses cases as you described, like client libraries. Though in such cases, if users somehow instantiate the models themselves rather than consuming from library methods, they won't get warnings. I guess this is beyond what we can do here since model initialisers are always public.

What about using something like "warning bits" for `mode`? For example, `mode=WarningMode.Get | WarningMode.Set` means warnings will be raised when the field is accessed (calling getter) and updated (calling setter). I was thinking that maybe we can simplify the cases to fields being set and fields being accessed and include model instantiation in the former, but this is possibly not enough for the client library case where warnings can be ignored at instantiation but should definitely be raised afterwards, so we might need at least one additional bit, something like `WarningMode.Create`. Expressing `mode` in these bits gives users full flexibility in a rather easy to understand and relate way.

**Comment by mzealey** [comment]
My use-case is to have deprecated show in the JSON schema, but still allow the field to be actively used. There's a work-around by setting `json_schema_extra={"deprecated": True}` but it's not particularly nice.

What'd I'd be looking for is a base model option (which we can then set in a base model inherited by all API interface classes) to allow this side-effect but nothing more?

**Comment by Trujifl** [comment]
Hi! I’ve been looking into point 2 of this issue: emitting a warning when a deprecated field is explicitly provided during model instantiation.

From what I understand, pydantic/pydantic#10865 attempted to implement this in `pydantic`, but was closed due to performance concerns. I also noticed that pydantic/pydantic-core#1551 moved the logic into `pydantic-core` and seemed to address the performance issue, but the discussion then shifted to the Python API/design question: whether warnings on instantiation should be unconditional, opt-in, or controlled by a more explicit deprecation mode.

Given that `pydantic-core` is now archived and appears to live in-tree in the main `pydantic` repository, I wanted to ask what the preferred direction is for this feature today.

Would maintainers still be interested in revisiting this, possibly with an explicit API for controlling when deprecated field warnings are emitted?

**Comment by flying-sheep** [comment]
That would be awesome! And while we’re at it, also if a FutureWarning or DeprecationWarning should be emitted.
